pkg/monitoring/vms/prometheus/prometheus.go

```diff
@@ -0,0 +1,276 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2018 Red Hat, Inc.
+ *
+ */
+
+package prometheus
+
+import (
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"kubevirt.io/kubevirt/pkg/log"
+	"kubevirt.io/kubevirt/pkg/version"
+	cmdclient "kubevirt.io/kubevirt/pkg/virt-handler/cmd-client"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/stats"
+)
+
+var (
+	// see https://www.robustperception.io/exposing-the-software-version-to-prometheus
+	versionDesc = prometheus.NewDesc(
+		"kubevirt_info",
+		"Version information",
+		[]string{"goversion", "kubeversion"},
+		nil,
+	)
+
+	storageIopsDesc = prometheus.NewDesc(
+		"kubevirt_vm_storage_iops",
+		"I/O operation performed.",
+		[]string{"domain", "drive", "type"},
+		nil,
+	)
+	// from now on: TODO: validate
+	vcpuUsageDesc = prometheus.NewDesc(
+		"kubevirt_vm_vcpu_time",
+		"Vcpu elapsed time, seconds.",
+		[]string{"domain", "id", "state"},
+		nil,
+	)
+	networkTrafficDesc = prometheus.NewDesc(
+		"kubevirt_vm_network_traffic_bytes",
+		"network traffic, bytes.",
+		[]string{"domain", "interface", "type"},
+		nil,
+	)
+	memoryUsageDesc = prometheus.NewDesc(
+		"kubevirt_vm_memory_amount_bytes",
+		"memory amount, bytes.",
+		[]string{"domain", "type"},
+		nil,
+	)
+)
+
+func Update(cli cmdclient.LauncherClient, ch chan<- prometheus.Metric) error {
+	vmStats, exists, err := cli.GetDomainStats()
+	if err != nil {
+		return err
+	}
+	if !exists || vmStats.Name == "" {
+		return nil
+	}
+
+	updateMemory(vmStats, ch)
+	updateVcpu(vmStats, ch)
+	updateBlock(vmStats, ch)
+	updateNetwork(vmStats, ch)
+
+	return nil
+}
+
+func updateMemory(vmStats *stats.DomainStats, ch chan<- prometheus.Metric) {
+	if vmStats.Memory.UnusedSet {
+		mv, err := prometheus.NewConstMetric(
+			memoryUsageDesc, prometheus.GaugeValue,
+			float64(vmStats.Memory.Unused),
+			vmStats.Name, "unused",
+		)
+		if err == nil {
+			ch <- mv
+		}
+	}
+	if vmStats.Memory.AvailableSet {
```

Comment:
checking if the values are "set" means that we update for each value at a time or we send a set of values if more than one is available ?

Reply:
Short answer: we will update all the metrics libvirt gives us, which can be a subset of all the supported metrics.

Long answer: libvirt bulk stats adhere to this tenet: always return all the _available_ data. So _each_  libvirt bulk stats call can report a subset of the supported metrics, and this is fine. Libvirt scrapes everything it gets, and return partial answers - better then no answer, that was the idea.

IOW, on error libvirts report no data, so we will have a hole in the metrics time serie.

Now, we the best we can do is bubble up the "no data", thus avoiding to report data we don't have.